eventsourcing/domain/model/aggregate.py

```diff
@@ -42,7 +42,8 @@ def __save__(self):
         except IndexError:
             pass
         if batch_of_events:
-            self.__publish_to_subscribers__(batch_of_events)
```

Comment:
Thanks for spending the time doing this @vladimirnani.

If we were to publish each of the pending events of an aggregate individually, rather than altogether as a list, they would need to be persisted independently, which would break the "atomic" quality of the AggregateRoot "save" method. The save method is intended to provide an "atomic" quality to the persisting of the events that result from a command, which is important for maintaining the coherence of the aggregate's sequence of records.

As I'm sure you understand, when one command results in many events, it can be important to prevent others picking up only some of the events, because it may represent an unworkable state, something nobody intended. Also, if another starts writing events after some of them have been persisted, but before all of them have been, the sequence of records would effectively be corrupted. The first command should fail, but events will have been recorded. The save method solves that problem by publishing the pending events to subscribers as a single list, so that they can be processed as a list, which means they can be persisted altogether, atomically, in the same database transaction. It also helps performance, but that's incidental.

The AggregateRoot class was added relatively recently, in order to provide this atomic quality. The super classes all publish events immediately and individually. When I added the AggregateRoot class, I did try to change all the subscribers in the library to handle lists of events, but you discovered I missed something: the subscribe_to decorator wasn't updated to support lists of events....

So rather than adjusting the AggregateRoot class, it might be better to adjust the decorator? Perhaps something like the following, which I basically copied from the PersistencePolicy class, would also fix the subscribe_to decorator?

```
    def event_type_predicate(event):
        if isinstance(event, (list, tuple)):
            return all(map(self.is_event, event))
        return self.event_type is None or isinstance(event, self.event_type)
```

Thanks again for opening this PR. If you wanted to adjust it, so we fix ``event_type_predicate()``, I would definitely merge it in.

Reply:
Thanks for the quick response. 
Agree with your point about atomic save. 

I have changed `publish` instead of predicate because we need to validate each event in batch before handling. In this case if you have a batch of events and one of which is `SomethingHappend` we would dispatch it in corresponding handler.